File: Code to Port/ClassicLib/ScanLog/SuspectScanner.py
"""
Suspect scanner module for CLASSIC.

This module scans for known crash patterns and suspects including:
- Checking main errors against known patterns
- Scanning call stacks for problematic signatures
- Identifying DLL-related crashes
- Matching against YAML-defined suspect patterns
"""

from typing import TYPE_CHECKING

from ClassicLib.ScanLog.ScanLogInfo import ClassicScanLogsInfo
from ClassicLib.Util import append_or_extend

if TYPE_CHECKING:
    from ClassicLib.ScanLog.ScanLogInfo import ClassicScanLogsInfo


class SuspectScanner:
    """Handles scanning for known crash patterns and suspects."""

    def __init__(self, yamldata: "ClassicScanLogsInfo") -> None:
        """
        Initialize the suspect scanner.

        Args:
            yamldata: Configuration data containing suspect patterns
        """
        self.yamldata: ClassicScanLogsInfo = yamldata

    def suspect_scan_mainerror(self, autoscan_report: list[str], crashlog_mainerror: str, max_warn_length: int) -> bool:
        """
        Scans the crash log for errors listed in a predefined suspect error list, updates the
        autoscan report upon detection, and determines if any suspects are found.

        Parameters:
        autoscan_report (list[str]): A list to store formatted strings of identified suspect
            errors and their associated details.
        crashlog_mainerror (str): The main error output from a crash log to scan for
            suspect errors.
        max_warn_length (int): The maximum length for formatting the error name in the autoscan
            report.

        Returns:
        bool: A boolean indicating whether any suspect errors were found in the crash log.
        """
        found_suspect = False

        for error_key, signal in self.yamldata.suspects_error_list.items():
            # Skip checking if signal not in crash log
            if signal not in crashlog_mainerror:
                continue

            # Parse error information
            error_severity, error_name = error_key.split(" | ", 1)

            # Format the error name for report
            formatted_error_name: str = error_name.ljust(max_warn_length, ".")

            # Add the error to the report
            report_entry: str = f"# Checking for {formatted_error_name} SUSPECT FOUND! > Severity : {error_severity} # \n-----\n"
            append_or_extend(report_entry, autoscan_report)

            # Update suspect found status
            found_suspect = True

        return found_suspect

    def suspect_scan_stack(
        self, crashlog_mainerror: str, segment_callstack_intact: str, autoscan_report: list[str], max_warn_length: int
    ) -> bool:
        """
        Analyzes a crash report and call stack information to identify potential suspect errors
        and integrates findings into an autoscan report. The function evaluates signal
        criteria from a YAML configuration to detect mismatches, potential issues, and specific
        patterns in the provided crash log and call stack details.

        Parameters:
            crashlog_mainerror (str): The main error extracted from the crash log.
            segment_callstack_intact (str): The intact segment of the call stack relevant to the analysis.
            autoscan_report (list[str]): A mutable report list where detected suspects are appended.
            max_warn_length (int): Maximum allowed length for warnings included in the report.

        Returns:
            bool: Indicates whether any suspect has been identified and added to the autoscan report.
        """
        any_suspect_found = False

        for error_key, signal_list in self.yamldata.suspects_stack_list.items():
            # Parse error information
            error_severity, error_name = error_key.split(" | ", 1)

            # Initialize match status tracking dictionary
            match_status = {
                "has_required_item": False,
                "error_req_found": False,
                "error_opt_found": False,
                "stack_found": False,
            }

            # Process each signal in the list
            should_skip_error = False
            for signal in signal_list:
                # Process the signal and update match_status accordingly
                if self._process_signal(signal, crashlog_mainerror, segment_callstack_intact, match_status):
                    should_skip_error = True
                    break

            # Skip this error if a condition indicates we should
            if should_skip_error:
                continue

            # Determine if we have a match based on the processed signals
            if self._is_suspect_match(match_status):
                # Add the suspect to the report and update the found status
                self._add_suspect_to_report(error_name, error_severity, max_warn_length, autoscan_report)
                any_suspect_found = True

        return any_suspect_found

    @staticmethod
    def _process_signal(signal: str, crashlog_mainerror: str, segment_callstack_intact: str, match_status: dict[str, bool]) -> bool:
        """
        Process an individual signal and update match status.

        Returns:
            True if processing should stop (NOT condition met)
        """
        # Constants for signal modifiers
        main_error_required = "ME-REQ"
        main_error_optional = "ME-OPT"
        callstack_negative = "NOT"

        if "|" not in signal:
            # Simple case: direct string match in callstack
            if signal in segment_callstack_intact:
                match_status["stack_found"] = True
            return False

        signal_modifier, signal_string = signal.split("|", 1)

        # Process based on signal modifier
        if signal_modifier == main_error_required:
            match_status["has_required_item"] = True
            if signal_string in crashlog_mainerror:
                match_status["error_req_found"] = True
        elif signal_modifier == main_error_optional:
            if signal_string in crashlog_mainerror:
                match_status["error_opt_found"] = True
        elif signal_modifier == callstack_negative:
            # Return True to break out of the loop if NOT condition is met
            return signal_string in segment_callstack_intact
        elif signal_modifier.isdecimal():
            # Check for minimum occurrences
            min_occurrences = int(signal_modifier)
            if segment_callstack_intact.count(signal_string) >= min_occurrences:
                match_status["stack_found"] = True

        return False

    @staticmethod
    def _is_suspect_match(match_status: dict[str, bool]) -> bool:
        """Determine if current error conditions constitute a suspect match."""
        if match_status["has_required_item"]:
            return match_status["error_req_found"]
        return match_status["error_opt_found"] or match_status["stack_found"]

    @staticmethod
    def _add_suspect_to_report(error_name: str, error_severity: str, max_warn_length: int, autoscan_report: list[str]) -> None:
        """Add a found suspect to the report with proper formatting."""
        formatted_error_name: str = error_name.ljust(max_warn_length, ".")
        message: str = f"# Checking for {formatted_error_name} SUSPECT FOUND! > Severity : {error_severity} # \n-----\n"
        append_or_extend(message, autoscan_report)

    @staticmethod
    def check_dll_crash(crashlog_mainerror: str, autoscan_report: list[str]) -> None:
        """
        A static method designed to analyze a crash log and identify if a DLL file is implicated in the crash.
        The method evaluates the primary error message from the crash log to determine if it references a DLL
        file, excluding specific cases such as "tbbmalloc". If a potentially problematic DLL is detected, the
        method appends a formatted notification to the autoscan report.

        Arguments:
            crashlog_mainerror: str
                The main error message extracted from the crash log, which is inspected for DLL-related
                mentions.
            autoscan_report: list[str]
                A reference to a list where any relevant findings or alerts about the crash will be appended.

        Returns:
            None
        """
        crashlog_mainerror_lower: str = crashlog_mainerror.lower()
        if ".dll" in crashlog_mainerror_lower and "tbbmalloc" not in crashlog_mainerror_lower:
            append_or_extend(
                (
                    "* NOTICE : MAIN ERROR REPORTS THAT A DLL FILE WAS INVOLVED IN THIS CRASH! * \n",
                    "If that dll file belongs to a mod, that mod is a prime suspect for the crash. \n-----\n",
                ),
                autoscan_report,
            )

File: Scanner111.Core/Analyzers/SuspectScanner.cs
using Microsoft.Extensions.Logging;
using Scanner111.Core.Infrastructure;
using Scanner111.Core.Models;
using Scanner111.Core.Models.Yaml;

namespace Scanner111.Core.Analyzers;

/// <summary>
/// Represents an analyzer that identifies known crash patterns and suspects from crash logs.
/// </summary>
/// <remarks>
/// Implements the IAnalyzer interface. Utilizes YAML settings for configuration and supports asynchronous analysis.
/// </remarks>
public class SuspectScanner : IAnalyzer
{
    private readonly IYamlSettingsProvider _yamlSettings;
    private readonly ILogger<SuspectScanner> _logger;

    /// <summary>
    ///     Initialize the suspect scanner
    /// </summary>
    /// <param name="yamlSettings">YAML settings provider for configuration</param>
    /// <param name="logger">Logger for internal diagnostics</param>
    public SuspectScanner(IYamlSettingsProvider yamlSettings, ILogger<SuspectScanner> logger)
    {
        _yamlSettings = yamlSettings;
        _logger = logger;
    }

    /// <summary>
    ///     Name of the analyzer
    /// </summary>
    public string Name => "Suspect Scanner";

    /// <summary>
    ///     Priority of the analyzer (lower values run first)
    /// </summary>
    public int Priority => 2;

    /// <summary>
    ///     Whether this analyzer can be run in parallel with others
    /// </summary>
    public bool CanRunInParallel => true;

    /// <summary>
    /// Analyze a crash log for suspect patterns
    /// </summary>
    /// <param name="crashLog">Crash log to analyze</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>Suspect analysis result</returns>
    public async Task<AnalysisResult> AnalyzeAsync(CrashLog crashLog, CancellationToken cancellationToken = default)
    {
        await Task.CompletedTask.ConfigureAwait(false); // Make it async-ready

        var reportLines = new List<string>();
        var errorMatches = new List<string>();
        var stackMatches = new List<string>();
        var matchDescriptions = new List<string>();

        const int maxWarnLength = 25;

        // Check for DLL crashes first
        CheckDllCrash(crashLog.MainError, reportLines);

        // Scan main error for suspects
        var mainErrorFound = SuspectScanMainError(reportLines, crashLog.MainError, maxWarnLength);
        if (mainErrorFound) errorMatches.Add(crashLog.MainError);

        // Scan call stack for suspects
        var callStackIntact = string.Join("\n", crashLog.CallStack);
        _logger.LogDebug(
            "Call stack has {CallStackCount} lines, joined length: {CallStackLength}", 
            crashLog.CallStack.Count, callStackIntact.Length);
        var stackFound = SuspectScanStack(crashLog.MainError, callStackIntact, reportLines, maxWarnLength);
        if (stackFound) stackMatches.Add(callStackIntact);

        return new SuspectAnalysisResult
        {
            AnalyzerName = Name,
            ErrorMatches = errorMatches,
            StackMatches = stackMatches,
            MatchDescriptions = matchDescriptions,
            ReportLines = reportLines,
            HasFindings = mainErrorFound || stackFound
        };
    }

    /// <summary>
    /// Scans the crash log for errors listed in a predefined suspect error list.
    /// Direct port of Python suspect_scan_mainerror method.
    /// </summary>
    /// <param name="autoscanReport">
    /// A list to store formatted strings of identified suspect errors and their associated details.
    /// </param>
    /// <param name="crashlogMainError">
    /// The main error output from a crash log to scan for suspect errors.
    /// </param>
    /// <param name="maxWarnLength">
    /// The maximum length for formatting the error name in the autoscan report.
    /// </param>
    /// <returns>
    /// A boolean indicating whether any suspect errors were found in the crash log.
    /// </returns>
    private bool SuspectScanMainError(List<string> autoscanReport, string crashlogMainError, int maxWarnLength)
    {
        var foundSuspect = false;

        // Load the strongly-typed YAML model
        var fallout4Yaml = _yamlSettings.LoadYaml<ClassicFallout4YamlV2>("CLASSIC Fallout4");
        if (fallout4Yaml?.CrashlogErrorCheck == null)
        {
            _logger.LogDebug("Could not load YAML or CrashlogErrorCheck section not found");
            return false;
        }

        // Parse the special format: "5 | Stack Overflow Crash: EXCEPTION_STACK_OVERFLOW"
        var suspectsErrorList = new Dictionary<string, (string severity, string criteria)>();
        foreach (var (key, value) in fallout4Yaml.CrashlogErrorCheck)
        {
            // Parse the key format: "5 | Stack Overflow Crash"
            var parts = key.Split(" | ", 2);
            if (parts.Length != 2) continue;
            var severity = parts[0].Trim();
            var description = parts[1].Trim();

            // Use description as key and store both severity and criteria
            suspectsErrorList[description] = (severity, value);
        }

        _logger.LogDebug("Loaded {SuspectCount} suspects from YAML", suspectsErrorList.Count);
        foreach (var (errorName, (severity, signal)) in suspectsErrorList)
        {
            // Skip checking if signal not in crash log
            if (!crashlogMainError.Contains(signal)) continue;

            // Format the error name for report
            var formattedErrorName = errorName.PadRight(maxWarnLength, '.');

            // Add the error to the report
            var reportEntry = $"# Checking for {formattedErrorName} SUSPECT FOUND! > Severity : {severity} # \n-----\n";
            autoscanReport.Add(reportEntry);

            // Update suspect found status
            foundSuspect = true;
        }

        return foundSuspect;
    }

    /// <summary>
    /// Analyzes the call stack of a crash log to identify potential suspects based on predefined patterns.
    /// </summary>
    /// <param name="crashlogMainError">The main error string extracted from the crash log.</param>
    /// <param name="segmentCallstackIntact">The complete concatenated call stack extracted from the crash log.</param>
    /// <param name="autoscanReport">A list to store the auto-scan report lines generated during the analysis.</param>
    /// <param name="maxWarnLength">The maximum length of warning messages to include in the report.</param>
    /// <returns>
    /// A boolean value indicating whether any suspect patterns were found in the call stack.
    /// </returns>
    private bool SuspectScanStack(string crashlogMainError, string segmentCallstackIntact, List<string> autoscanReport,
        int maxWarnLength)
    {
        var anySuspectFound = false;

        // Load the strongly-typed YAML model
        var fallout4Yaml = _yamlSettings.LoadYaml<ClassicFallout4YamlV2>("CLASSIC Fallout4");
        if (fallout4Yaml?.CrashlogStackCheck == null) return false;

        // Use the already properly typed CrashlogStackCheck
        var suspectsStackList = fallout4Yaml.CrashlogStackCheck;

        _logger.LogDebug("Found {StackPatternCount} stack patterns to check", suspectsStackList.Count);
        _logger.LogDebug(
            "Call stack preview: '{CallStackPreview}'", 
            segmentCallstackIntact.Substring(0, Math.Min(200, segmentCallstackIntact.Length)));
        foreach (var (errorKey, signalList) in suspectsStackList)
        {
            // Parse error information from the key format: "6 | BA2 Limit Crash"
            var keyStr = errorKey;
            _logger.LogDebug("Processing stack pattern: '{PatternKey}' with {SignalCount} signals", keyStr, signalList.Count);
            var parts = keyStr.Split(" | ", 2);
            if (parts.Length < 2)
            {
                _logger.LogDebug("Skipping malformed stack pattern key: '{PatternKey}'", keyStr);
                continue;
            }

            var errorSeverity = parts[0];
            var errorName = parts[1];

            // Initialize match status tracking dictionary
            var matchStatus = new Dictionary<string, bool>
            {
                { "has_required_item", false },
                { "error_req_found", false },
                { "error_opt_found", false },
                { "stack_found", false }
            };

            // Process each signal in the list
            var shouldSkipError = false;
            foreach (var signal in signalList)
            {
                _logger.LogDebug("Processing signal '{Signal}' for pattern '{ErrorName}'", signal, errorName);
                // Process the signal and update match_status accordingly
                if (!ProcessSignal(signal, crashlogMainError, segmentCallstackIntact, matchStatus)) continue;
                _logger.LogDebug("Signal '{Signal}' triggered skip (NOT condition met)", signal);
                shouldSkipError = true;
                break;
            }

            // Skip this error if a condition indicates we should
            if (shouldSkipError) continue;

            // Determine if we have a match based on the processed signals
            if (!IsSuspectMatch(matchStatus)) continue;
            // Add the suspect to the report and update the found status
            AddSuspectToReport(errorName, errorSeverity, maxWarnLength, autoscanReport);
            anySuspectFound = true;
        }

        return anySuspectFound;
    }


    /// <summary>
    /// Processes a signal to analyze matching conditions in crash logs and updates the match status accordingly.
    /// </summary>
    /// <param name="signal">The signal string to be processed, which may include a modifier and a target string separated by '|'.</param>
    /// <param name="crashlogMainError">The main error section of the crash log to be checked against the signal.</param>
    /// <param name="segmentCallstackIntact">The callstack segment of the crash log to be examined for signal matches.</param>
    /// <param name="matchStatus">A dictionary for tracking various match conditions and their statuses.</param>
    /// <returns>
    /// A boolean value indicating whether a "NOT" condition is met in the signal, triggering a skip in further processing.
    /// </returns>
    private bool ProcessSignal(string signal, string crashlogMainError, string segmentCallstackIntact,
        Dictionary<string, bool> matchStatus)
    {
        // Constants for signal modifiers
        const string mainErrorRequired = "ME-REQ";
        const string mainErrorOptional = "ME-OPT";
        const string callstackNegative = "NOT";

        if (!signal.Contains("|"))
        {
            // Simple case: direct string match in callstack
            _logger.LogDebug("Simple signal check: '{Signal}' in call stack", signal);
            if (segmentCallstackIntact.Contains(signal))
            {
                _logger.LogDebug("FOUND simple signal '{Signal}' in call stack!", signal);
                matchStatus["stack_found"] = true;
            }
            else
            {
                _logger.LogDebug("Simple signal '{Signal}' NOT found in call stack", signal);
            }

            return false;
        }

        var signalParts = signal.Split('|', 2);
        var signalModifier = signalParts[0];
        var signalString = signalParts[1];

        switch (signalModifier)
        {
            // Process based on signal modifier
            case mainErrorRequired:
            {
                matchStatus["has_required_item"] = true;
                if (crashlogMainError.Contains(signalString)) matchStatus["error_req_found"] = true;
                break;
            }
            case mainErrorOptional:
            {
                if (crashlogMainError.Contains(signalString)) matchStatus["error_opt_found"] = true;
                break;
            }
            case callstackNegative:
                // Return True to break out of the loop if NOT condition is met
                return segmentCallstackIntact.Contains(signalString);
            default:
            {
                if (int.TryParse(signalModifier, out var minOccurrences))
                {
                    // Check for minimum occurrences
                    var occurrences = CountOccurrences(segmentCallstackIntact, signalString);
                    if (occurrences >= minOccurrences) matchStatus["stack_found"] = true;
                }

                break;
            }
        }

        return false;
    }

    /// <summary>
    /// Counts the occurrences of a specific substring within a given text.
    /// </summary>
    /// <param name="text">The text in which to search for the substring.</param>
    /// <param name="pattern">The substring to count within the text.</param>
    /// <returns>The total number of occurrences of the substring in the text.</returns>
    private static int CountOccurrences(string text, string pattern)
    {
        if (string.IsNullOrEmpty(pattern))
            return 0;

        var count = 0;
        var index = 0;
        while ((index = text.IndexOf(pattern, index, StringComparison.Ordinal)) != -1)
        {
            count++;
            index += pattern.Length;
        }

        return count;
    }


    /// <summary>
    /// Determines if a suspect matches based on the provided match status.
    /// </summary>
    /// <param name="matchStatus">A dictionary containing the match status of various conditions.</param>
    /// <returns>
    /// Returns <c>true</c> if the suspect match criteria are satisfied; otherwise, returns <c>false</c>.
    /// </returns>
    private static bool IsSuspectMatch(Dictionary<string, bool> matchStatus)
    {
        if (matchStatus["has_required_item"]) return matchStatus["error_req_found"];
        return matchStatus["error_opt_found"] || matchStatus["stack_found"];
    }


    /// <summary>
    /// Adds a suspect to the report based on the provided error details.
    /// </summary>
    /// <param name="errorName">The name of the error or suspect identified.</param>
    /// <param name="errorSeverity">The severity level of the identified error.</param>
    /// <param name="maxWarnLength">The maximum allowed length for formatting warnings.</param>
    /// <param name="autoscanReport">The report to which the formatted suspect details should be added.</param>
    private static void AddSuspectToReport(string errorName, string errorSeverity, int maxWarnLength,
        List<string> autoscanReport)
    {
        var formattedErrorName = errorName.PadRight(maxWarnLength, '.');
        var message = $"# Checking for {formattedErrorName} SUSPECT FOUND! > Severity : {errorSeverity} # \n-----\n";
        autoscanReport.Add(message);
    }


    /// <summary>
    /// Checks if the main error of the crash log indicates involvement of a DLL file and updates the report accordingly.
    /// </summary>
    /// <param name="crashlogMainError">The main error message from the crash log.</param>
    /// <param name="autoscanReport">A list for appending the report lines generated during the check.</param>
    private static void CheckDllCrash(string crashlogMainError, List<string> autoscanReport)
    {
        var crashlogMainErrorLower = crashlogMainError.ToLower();
        if (crashlogMainErrorLower.Contains(".dll") && !crashlogMainErrorLower.Contains("tbbmalloc"))
            autoscanReport.AddRange([
              "* NOTICE : MAIN ERROR REPORTS THAT A DLL FILE WAS INVOLVED IN THIS CRASH! * \n",
                "If that dll file belongs to a mod, that mod is a prime suspect for the crash. \n-----\n"
            ]);
    }
}